### DISCUSSION TITLE
pb_hooks transactions - create account and insert related records?

### DISCUSSION THREAD

**Discussion by benallfree** [pocketbase/pocketbase#3670] [Q&A] [answered]
I've looked at the pb_hooks API but it's still somewhat unclear - how would I create a new auth record and insert related records inside a transaction, and rollback with a JSON error message if anything fails? Will try/catch/throw work? I'd like to do something like this:

```
// GET /signup

const instanceName = generateInstanceName()
res.json( { instanceName } )  

// POST /signup
try {
   // create user account
} catch (e) {
   res.json({ status: 'error', message: e.toString() })
   throw new Error(`Rollback`)
}

try {
   // insert an instance record
} catch (e) {
   const instanceName = generateNewInstanceName()
   res.json({ status: 'error', message: e.toString(), restart: { instanceName })
   throw new Error(`Rollback`)
}
```

**Comment by ganigeorgiev** [comment, marked as answer]
You need to wrap your call in a `$app.dao().runInTransaction(...)` and use the transaction dao when performing the db operations. Something like:

```js
const usersCollection = $app.dao().findCollectionByNameOrId("users")
const exampleCollection = $app.dao().findCollectionByNameOrId("example");

try {
    // see https://pocketbase.io/docs/js-records/#transaction
    $app.dao().runInTransaction((txDao) => {
        const user = new Record(usersCollection);
        user.set("name", "John Doe");
        user.set("email", "test@example.com");
        user.set("username", "u_" + $security.randomStringWithAlphabet(5, "123456789"))
        user.setPassword("1234567890");
        txDao.saveRecord(user);

        const foo = new Record(exampleCollection)
        foo.set("title", "something");
        txDao.saveRecord(foo);
    });
} catch (err) {
    // see https://pocketbase.io/docs/js-routing/#error-response
    throw new BadRequestError("Something went wrong...", err);
}
```

_Edit: updated the example to show creating an auth record._

**Reply by benallfree** [reply]
That looks about right... thank you!

**Reply by benallfree** [reply]
Is that the official way to create a user record? Do I need to generate the `user1234` username manually?

**Reply by benallfree** [reply]
I'm having a little trouble understanding how to cancel a transaction and return a custom error payload. Suppose I want to send

```
c.json(500, {
            fields: {
              email: `Could not create user: ${e}`,
            },
          })
```

...but not necessarily throw an exception. Is that supported, or am I required to throw an exception to rollback a transaction?

**Reply by ganigeorgiev** [reply]
>  Do I need to generate the user1234 username manually?

Sorry, I forgot about the username because it is automatically set when using the upsert form. When using directly the Dao you need to set it manually. To autogenerate it, you can use something like:
```js
$app.dao().suggestUniqueAuthRecordUsername("users", "u_" + $security.randomStringWithAlphabet(5, "123456789"))
```
(_the `suggestUniqueAuthRecordUsername` is optional and will check for existence and will append random numbers in case of duplication; with the planned refactoring I'll consider unifying the persistence handling when using a form or directly the Dao_)

> I'm having a little trouble understanding how to cancel a transaction...

You can rollback a transaction manually by calling `txDao.db().rollback()` but I don't think it is necessarily in your case because throwing an error inside the transaction callback should work.
You can do anything in the outer `catch` block (_the `BadRequestError` rethrow was just an example_).

**Reply by benallfree** [reply]
Here's the final handler. Do you see anything that could be done in a more PocketBase-esque way? :) 

```js
routerAdd(
  'POST',
  '/api/signup',
  (c) => {
    const parsed = (() => {
      const rawBody = readerToString(c.request().body)
      try {
        const parsed = JSON.parse(rawBody)
        return parsed
      } catch (e) {
        throw new BadRequestError(
          `Error parsing payload. You call this JSON? ${rawBody}`,
          e,
        )
      }
    })()
    const email = parsed.email?.trim()
    const password = parsed.password?.trim()
    const desiredInstanceName = parsed.instanceName?.trim()

    const error = (fieldName, slug, description, extra) =>
      new ApiError(500, description, {
        [fieldName]: new ValidationError(slug, description),
        ...extra,
      })

    if (!email) {
      throw error(`email`, 'required', 'Email is required')
    }

    if (!password) {
      throw error(`password`, `required`, 'Password is required')
    }

    if (!desiredInstanceName) {
      throw error(`instanceName`, `required`, `Instance name is required`)
    }

    const userExists = (() => {
      try {
        const record = $app.dao().findFirstRecordByData('users', 'email', email)
        return true
      } catch {
        return false
      }
    })()

    if (userExists) {
      throw error(
        `email`,
        `exists`,
        `That user account already exists. Try a password reset.`,
      )
    }

    $app.dao().runInTransaction((txDao) => {
      const usersCollection = $app.dao().findCollectionByNameOrId('users')
      const instanceCollection = $app
        .dao()
        .findCollectionByNameOrId('instances')

      const user = new Record(usersCollection)
      try {
        const username = $app
          .dao()
          .suggestUniqueAuthRecordUsername(
            'users',
            'user' + $security.randomStringWithAlphabet(5, '123456789'),
          )
        user.set('username', username)
        user.set('email', email)
        user.setPassword(password)
        txDao.saveRecord(user)
      } catch (e) {
        throw error(`email`, `fail`, `Could not create user: ${e}`)
      }

      try {
        const instance = new Record(instanceCollection)
        instance.set('subdomain', desiredInstanceName)
        instance.set('uid', user.get('id'))
        instance.set('status', 'Idle')
        instance.set('version', '~0.19.0')
        txDao.saveRecord(instance)
      } catch (e) {
        if (e.toString().match(/ UNIQUE /)) {
          throw error(
            `instanceName`,
            `exists`,
            `Instance name was taken, sorry aboout that. Try another.`,
          )
        }
        throw error(`instanceName`, `fail`, `Could not create instance: ${e}`)
      }

      $mails.sendRecordVerification($app, user)
    })

    return c.json(200, { status: 'ok' })
  } /* optional middlewares */,
)
```

**Reply by ganigeorgiev** [reply]
>  will inserting a record in the users collection send out a verification email or how would I do that manually?

No, it will just insert the record in the related collection. It seems that I have forgotten to document the types for it, but there is a `$mails` binding that you can use in similar manner to the [same Go package](https://pkg.go.dev/github.com/pocketbase/pocketbase/mails):

```js
$mails.sendRecordVerification($app, user)
```

--- 

In general the above code looks ok to me. Couple things eventually you can change depending on the desired behavior and personal preferences:

- For other ways to read request body parameters you can check https://pocketbase.io/docs/js-routing/#reading-request-body.
    ```js
    const data = new DynamicModel({
        // describe the shape of the fields to read (used also as initial values)
        email:        "",
        password:     "",
        instanceName: "",
    })

    c.bind(data)

    console.log(data.email)
    ```

- If you are creating the records from untrusted user input and want to trigger the builtin validations, you can use the [`RecordUpsertForm`](https://pocketbase.io/docs/js-records/#create-new-record-with-data-validations). The only caveat in your case since you need to execute it in a transaction is that you have to explicitly set its dao using `form.SetDao(txDao)`. Here is an example creating an auth record:
    ```js
    const collection = $app.dao().findCollectionByNameOrId("users")

    const record = new Record(collection)

    const form = new RecordUpsertForm($app, record)

    // use the transactional dao
    form.setDao(txDao);

    // or you can also use directly `form.loadRequest(request, "")`
    // to skip manually parsing the body when the submitted body fields match with the record ones
    form.loadData({
        "email":           email,
        "password":        password,
        "passwordConfirm": password,
        // you can skip this since the form will autogenerate it for you
        // "username": username,
    })

    // validates the loaded data and create the record
    try {
        form.submit();
    } catch (err) {
        throw new BadRequestError("Failed to create user.", err) // the err in this case will be formatted as it is map of ValidationError
    }
    ```
    
---

_Side-note: The existing API warts (and the availability of multiple ways to do pretty much the same thing) will be refactored to simplify the exposed internals and reduce the ambiguity but that will be after Presentator. It will probably be a breaking change but I think it will be better both in terms of maintenance and easy of use (there will be an official announcement once I get time to outline the planned changes)._

**Reply by benallfree** [reply]
https://pocketbase.io/docs/js-routing/#reading-request-body is for `form-data` not JSON though, right? If I'm expecting a JSON payload I need to parse it myself?

I think for now this function is about as tight as it can be. The signup form takes email/password/instance-name, but other fields need to be defined when actually creating the instance record and the `instances` table only allows admins to write to it.

**Reply by ganigeorgiev** [reply]
> https://pocketbase.io/docs/js-routing/#reading-request-body is for form-data not JSON though, right? If I'm expecting a JSON payload I need to parse it myself?

No, it should work with both. Only `c.formValue()`, `c.formFile()`, and other similar echo methods, they target `multipart/form-data` requests.

**Comment by benallfree** [comment]
@ganigeorgiev How would parsing nested JSON with optional fields work? Imagine this:

```ts
type Payload = {
   instanceId: string
   fields: Partial<{maintenance:boolean, name: string, version:string}>
}
```

```json
// Example payload
{
  "instanceId": "kz4ngg77eaw1ho0",
  "fields": {
    "maintenance": true
  }
```

```js
// index.pb.js

const data = new DynamicModel({
    instanceName: "",
    // ???
})
```

**Reply by ganigeorgiev** [reply]
A plain object should work, something like:

```js
const data = new DynamicModel({
    instanceName: "",
    fields: {}
})
```

**Reply by benallfree** [reply]
Destructuring is weird:

```js
    // This is necessary for destructuring to work correctly
    data = JSON.parse(JSON.stringify(data))

    const {
      instanceId,
      fields: { maintenance, name, version },
    } = data
```

**Reply by benallfree** [reply]
Sorry a couple more questions:

```js
/*
{
  "instanceId": "kz4ngg77eaw1ho0",
  "fields": {
    "maintenance": true
    "name": '',
    "version": ''
  }
}
*/
routerAdd(
  'PUT',
  '/api/instance',
  (c) => {
    console.log(`***TOP OF PUt`)
    let data = new DynamicModel({
      // describe the shape of the fields to read (used also as initial values)
      instanceId: '',
      fields: {
        maintenance: null,
        name: null,
        version: null,
      },
    })

    c.bind(data)

    // This is necessary for destructuring to work correctly
    data = JSON.parse(JSON.stringify(data))

    const {
      instanceId,
      fields: { maintenance, name, version },
    } = data

    console.log(
      `***vars`,
      JSON.stringify({ instanceId, maintenance, name, version }),
    )

    function cleanObject(obj) {
      return Object.entries(obj).reduce((acc, [key, value]) => {
        if (value !== null && value !== undefined) {
          acc[key] = value
        }
        return acc
      }, {})
    }

    console.log(`***original`, JSON.stringify(data))
    const sanitized = cleanObject({ subdomain: name, version, maintenance })
    console.log(`***sanitized`, JSON.stringify(sanitized))

    const record = $app.dao().findRecordById('instances', instanceId)

    const form = new RecordUpsertForm($app, record)
    form.loadData(sanitized)
    form.submit()

    return c.json(200, { status: 'ok' })
  } /* optional middlewares */,
)
```

1. Will DynamicModel filter out unknown keys? For example, if the user tried to pass an admin-only field in the payload?
2. My current API rules are set to allow updates for admins only. Why is this router path allowing the record update by a request originating from a regular user (not admin)? Is it bypassing the API rules?

**Reply by ganigeorgiev** [reply]
1. `c.bind(data)` should bind only the fields that are defined in the schema (_I'm not sure about the nested fields though since internally `fields` will be a Go map_).

2. Collection API rules are checked only when using their corresponding CRUD API route. If you have a custom route it is up to the developers to perform whatever access check they need. If you want your route to be accessible only by admins, you can pass after the route handler the `$apis.requireAdminAuth()` middleware (see also [Builtin middlewares](https://pocketbase.io/docs/js-routing/#builtin-middlewares)).
